### PR TITLE
Adding workflow for poetry deployment

### PR DESCRIPTION
**Pull request by andrea-pasquale**
In this PR I'm addressing this point:
> [ ] There is no workflow for poetry deployment in this repository and in the workflows.

From qiboteam/qibocal#249.
I will also open the corresponding PR in qibocal.
I usually don't play a lot with workflows so it may be possible that I've missed something.

### CODE REVIEW

File: .github/workflows/deploy-pip-poetry.yml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+
+jobs:
+  build:
+    runs-on: ${{ inputs.os }}
+    defaults:
+      run:
+        shell: bash
+    steps:
+      - uses: actions/checkout@v2
```
Comment (AleCandido) [inline .github/workflows/deploy-pip-poetry.yml:29]:
```suggestion
      - uses: actions/checkout@v3
```
https://github.com/actions/checkout

File: .github/workflows/deploy-pip-poetry.yml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+      - name: Setup python
+        uses: actions/setup-python@v2
```
Comment (AleCandido) [inline .github/workflows/deploy-pip-poetry.yml:31]:
```suggestion
        uses: actions/setup-python@v4
```
https://github.com/actions/setup-python

**Review by AleCandido — CHANGES_REQUESTED** [review]
I know I'm not doing myself in NNPDF, but you should use latest version available of canonical actions

**Comment by AleCandido** [comment]
Apart for versions it should be pretty much alright :)

**Comment by Edoardo-Pedicillo** [comment]
This workflow looks like [rules-poetry.yml ](https://github.com/qiboteam/workflows/blob/main/.github/workflows/rules-poetry.yml), the only difference is the publish job. I suggest you add an input in `rules-poetry.yml`, update it with the new job and change the file name, if necessary. What does @AleCandido think about it ?

**Comment by AleCandido** [comment]
> This workflow looks like [rules-poetry.yml ](https://github.com/qiboteam/workflows/blob/main/.github/workflows/rules-poetry.yml?rgh-link-date=2023-02-20T11%3A44%3A53Z), the only difference is the publish job. I suggest you add an input in `rules-poetry.yml`, update it with the new job and change the file name, if necessary. What does @AleCandido think about it ?

You're right it's an option, and having less workflows improves the maintenance w.r.t. many redundant objects. This is a clear advantage.

I'm not fully convinced because they are really performing two different tasks, so the perfect solution, in a regular programming language, would be to split the common part, and have two different functions that both call the common one, instead of branching inside a single one with a boolean parameter.
However, here function calls are nested workflows, so it's definitely not an ideal solution, and I would try to keep the final solution as plain as possible (and avoid too much nesting, but if possible branching as well).

So, maybe I'm slightly in favor of keeping this separate, but I acknowledge that this might be due to completely personal taste. Please, feel free to make your own considerations and pick your favorite solution.

**Comment by andrea-pasquale** [comment]
> This workflow looks like [rules-poetry.yml ](https://github.com/qiboteam/workflows/blob/main/.github/workflows/rules-poetry.yml), the only difference is the publish job. I suggest you add an input in `rules-poetry.yml`, update it with the new job and change the file name, if necessary. What does @AleCandido think about it ?

I also don't have a strong preference about this.
@scarrazza if you are fine with having two different workflows for tests and deployment this is ready to merge.

**Comment by scarrazza** [comment]
That is fine to me, thank you.

**Review by scarrazza — APPROVED** [review]

File: .github/workflows/deploy-pip-poetry.yml [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+  workflow_call:
+    inputs:
+        os:
+          required: true
+          type: string
+        python-version:
+          required: true
+          type: string
+        publish:
```
Comment (Edoardo-Pedicillo) [inline .github/workflows/deploy-pip-poetry.yml:13]: The `publish` input is not used in the workflow

Comment (andrea-pasquale) [reply .github/workflows/deploy-pip-poetry.yml:13]: I see. If it is not necessary I will remove it.

**Comment by andrea-pasquale** [comment]
@Edoardo-Pedicillo let me know if this is now fine to merge.

**Review by AleCandido — APPROVED** [review]

**Pull request merged**